### ISSUE TITLE
Error building version 4.3.0

### ISSUE THREAD

**Issue by myhro** [joeyespo/grip#196]
Hi Joe,

I was trying to update the grip Debian package and noticed that the build process is failing in this step:

    (...)
    creating build/lib.linux-x86_64-2.7/grip/static
    copying grip/static/favicon.ico -> build/lib.linux-x86_64-2.7/grip/static
    error: can't copy 'grip/static/octicons': doesn't exist or not a regular file

This can be reproduced using `python setup.py install` as well.

Looks like the [this line][1] is treating `grip/static/octions` as a file, not a subfolder. I've tried to work around this [with a function like][2]:

```python
def get_data_files():
    for d in ['static/', 'templates/', 'vendor/']:
        data_files = []
        for (path, dirs, files) in os.walk(os.path.join('grip/', d)):
            for f in files:
                data_files.append(os.path.join(path, f))
        return data_files
```

... And had no luck. The build didn't failed, but looks like octicons files aren't being added.

Do you have any idea how we can overcome this?

Regards,
Tiago.

[1]: https://github.com/joeyespo/grip/blob/v4.3.0/setup.py#L45
[2]: https://stackoverflow.com/a/36693250

**Comment by joeyespo** [comment]
Huh, that's odd. Could you try changing the single star to a double star? I
wonder if Debian or Python uses that glob style (I'll dig deeper later).
Thanks!

**Comment by myhro** [comment]
> Huh, that's odd. Could you try changing the single star to a double star?

Tried, but didn't worked as well.

> I wonder if Debian or Python uses that glob style (I'll dig deeper later).

Actually the build process under Debian fails when calling `setup.py install`.

I'm wondering how you were able to build and upload it to PyPI without facing this error. Anyway, looks like the [package in there][1] doesn't contains octicons files:

    $ 7z l grip-4.3.0.zip  | grep static
    2016-04-09 03:45:36 .....         1406          474  grip-4.3.0/grip/static/favicon.ico

Have you tested it?

[1]: https://pypi.python.org/packages/dd/6c/fae849704d4884351938befbc417c30df9d7661620afeff75c4dbb936c00/grip-4.3.0.zip#md5=9715211e0a92927cc7fb4cd60329cd58

**Comment by joeyespo** [comment]
I did test it, though with `pip install`, not `setup install` directly. I get runtime 404 errors instead of build errors. So looks like you're right. 

Looking into a solution now.

**Comment by joeyespo** [comment]
Ok. So different things happen between `setup.py sdist` and `setup.py bdist`. Some builds include files from `MANIFEST.in` and others from `package_data`. Then there's pre-2.7 behavior, distutils vs setuptools, a `package_data` bug, and the `include_package_data` approach (not currently used by Grip). *whew!*

Looks like [the best approach](http://stackoverflow.com/questions/2968701/2-techniques-for-including-files-in-a-python-distribution-which-is-better/2969087#2969087) is to then include non-Python files your project consumes through both `package_data` *and* `MANIFEST.in` (note [the comment](http://stackoverflow.com/questions/2968701/2-techniques-for-including-files-in-a-python-distribution-which-is-better/2969087#comment3084734_2969087) about the bug).

`MANIFEST.in` gives you the convenient `recursive-include` directive. `package_data` does not. So I'll have to remember that subdirectories need explicitly added. (I'm not going to try to be more clever than that right now, since I can't think of a good abstraction for `package_data` at the moment. I'd rather think through the pros and cons of a reusable implementation later on.) This is done in #198.

Here's a bunch of references:
- https://docs.python.org/2/distutils/sourcedist.html#manifest-template
- http://stackoverflow.com/questions/2968701/2-techniques-for-including-files-in-a-python-distribution-which-is-better/2969087#2969087
- http://stackoverflow.com/questions/27664504/how-to-add-package-data-recursively-in-python-setup-py/36693250#36693250
- http://stackoverflow.com/questions/7522250/how-to-include-package-data-with-setuptools-distribute
- http://stackoverflow.com/questions/1612733/including-non-python-files-with-setup-py

**Comment by joeyespo** [comment]
Ok. Try v4.3.1. Feel free to re-open if that didn't fix it for you.

**Comment by myhro** [comment]
Thank you very much for this careful research. Indeed, it solved the problem for PyPI:

    $ 7z l grip-4.3.1.zip | grep static
    2016-04-09 03:45:36 .....         1406          474  grip-4.3.1/grip/static/favicon.ico
    2016-07-19 01:44:28 .....         9525         1837  grip-4.3.1/grip/static/octicons/octicons.css
    2016-07-19 01:44:28 .....        44098        24032  grip-4.3.1/grip/static/octicons/octicons.eot
    2016-07-19 01:44:28 .....        45369        13706  grip-4.3.1/grip/static/octicons/octicons.svg
    2016-07-19 01:44:28 .....        43920        23942  grip-4.3.1/grip/static/octicons/octicons.ttf
    2016-07-19 01:44:28 .....        24004        23976  grip-4.3.1/grip/static/octicons/octicons.woff
    2016-07-19 01:44:28 .....        20248        20258  grip-4.3.1/grip/static/octicons/octicons.woff2

But I'm still facing errors when running `setup.py install` (on both Python 2 (2.7.9 and 2.7.12) and Python 3 (3.4.2 and 3.5.2)):

`error: can't copy 'grip/static/octicons': doesn't exist or not a regular file`

Looks like the first glob `'static/*'` is still including `octicons` as a file.

**Comment by joeyespo** [comment]
Interesting. Found another reference with what looks like the same problem:

- http://stackoverflow.com/questions/3712033/python-distutils-error-directory-doesnt-exist-or-not-a-regular-file

If this isn't a distutils bug, it's quite an annoying feature!

Ok. Try v4.3.2.

**Comment by myhro** [comment]
Absolutely awesome, Joe! Thanks again for digging so deep in this issue.

> If this isn't a distutils bug, it's quite an annoying feature!

I completely agree! :-)

> Ok. Try v4.3.2.

Can confirm that it is fixed in `v4.3.2`. I'll be closing this issue.
